test/util.js

```diff
@@ -513,3 +515,261 @@ function checkCursorPaginatedHistoryObject ({
+
+  const sortValues = (values) => {
+    return values.slice(0).sort((a, b) => {
+      if (a < b) return -1
+      else if (a > b) return 1
+      else return 0
+    })
+  }
+
+  // prop=value
+  const getExactValueFilter = (prop, exactValue) => `${prop}=${exactValue}`
+
+  // prop[]=value1&prop[]=value2
+  const getArrayValuesFilter = (prop, values) => values.map(v => `${prop}[]=${v}`).join('&')
+
+  // prop=value1,value2
+  const getArrayCommaSeparatedValuesFilter = (prop, values) => `${prop}=${values.join(',')}`
+
+  // prop[gte]=min&prop[lte]=max
+  const getRangeFilter = (prop, range) => `${prop}[gte]=${range.gte}&${prop}[lte]=${range.lte}`
+
+  const addCheckObjFunctions = (filter) => {
+    const {
+      prop,
+      customGetValue,
+      isArrayFilter,
+      isRangeFilter,
+      customCheck
+    } = filter
+
+    const getValue = (obj) => _.isFunction(customGetValue) ? customGetValue(obj) : obj[prop]
+    filter.getValue = getValue
+
+    filter.getExactValueCheckObj = _.curry((exactValue, obj) => {
+      return _.isFunction(customCheck)
+        ? customCheck(obj, exactValue)
+        : getValue(obj) === exactValue
+    })
+
+    if (isArrayFilter) {
+      filter.getArrayValuesCheckObj = _.curry((values, obj) => {
+        return _.isFunction(customCheck)
+          ? customCheck(obj, values)
+          : values.includes(getValue(obj))
+      })
+    } else if (isRangeFilter) {
+      filter.getRangeValuesCheckObj = _.curry((rangeValues, obj) => {
+        return _.isFunction(customCheck)
+          ? customCheck(obj, rangeValues)
+          : rangeValues.gte <= getValue(obj) && getValue(obj) <= rangeValues.lte
+      })
+    }
+  }
+
+  const addTestValues = (filter) => {
+    const {
+      isArrayFilter,
+      isRangeFilter,
+      customTestValues,
+    } = filter
+
+    addCheckObjFunctions(filter)
+
+    const getValue = filter.getValue
+
+    const useCustomValues = customTestValues && customTestValues.length
+
+    const realValues = _.compact(_.uniq(results.map(getValue)))
+    let testValues = useCustomValues
+      ? customTestValues
+      : _.shuffle(realValues).slice(0, 2) // only take few values
+
+    if (!testValues.length) testValues = ['value1', 'value2']
+
+    const sortedValues = sortValues(testValues)
+
+    filter.shouldHaveResults = !useCustomValues && Boolean(realValues.length)
+
+    if (useCustomValues) filter.testExactValues = testValues
+    else filter.testExactValues = [getRandomValue(testValues)]
+
+    if (isArrayFilter) filter.testArrayValues = testValues
+
+    if (isRangeFilter) {
+      if (testValues.length === 1) {
+        const randomValue = getRandomValue(sortedValues)
+
+        filter.testRangeValues = {
+          gte: randomValue,
+          lte: randomValue,
+        }
+      } else {
+        filter.testRangeValues = {
+          gte: _.first(sortedValues),
+
+          // filter only on the first half set of results to check if the range filter is effective
```

Comment:
Why?

Reply:
This is for the same reason as array filter. We filter only a subset of available values (half of the interval is arbitrary) to ensure the range filter works.

The condition `rangeValues.gte <= getValue(obj) && getValue(obj) <= rangeValues.lte` will check that the returned results have the expected values.
https://github.com/stelace/stelace/blob/47a66145b591f0c5e35d0dd3ac5c5be1107d1ba4/test/util.js#L608-L612

Reply:
> We filter only a subset of available values (half of the interval is arbitrary) to ensure the range filter works.

This still does not explain why. You’re merely rephrasing your current comment I was asking about.

And `rangeValues.gte <= getValue(obj) && getValue(obj) <= rangeValues.lte` looks like a typo to me (do you really want `<= lte`)?

Reply:
I'll add an example for clarity.
Let's suppose we have a number filter on resources whose values spreads from 1 to 10.
If we didn't pick a subset of the interval, the filter will apply the range `[1, 10]`, which is equivalent to the absence of the filter in this case because we filter all results. By filtering the range `[1, 5]` and checking that results match the filter, we ensure it works.
In the last version of the code, I picked random values (if there are more than one value) so it isn't half of the interval.
https://github.com/stelace/stelace/blob/3941883f5e2bf591034f4e4ac97693269ff44526/test/util.js#L679-L698

Concerning the condition `rangeValues.gte <= getValue(obj) && getValue(obj) <= rangeValues.lte`, it seems to me there isn't a typo. This function was refactored in a subsequent commit to allow all range combinations. Maybe it's clearer this way.
https://github.com/stelace/stelace/blob/3941883f5e2bf591034f4e4ac97693269ff44526/test/util.js#L600-L609

Reply:
> If we didn't pick a subset of the interval, the filter will apply the range [1, 10], which is equivalent to the absence of the filter in this case because we filter all results. By filtering the range [1, 5] and checking that results match the filter, we ensure it works.

This is the piece of simple explanation I was looking for, thanks.

> Concerning the condition `rangeValues.gte <= getValue(obj) && getValue(obj) <= rangeValues.lte`, it seems to me there isn't a typo.

My bad, I’ve read too fast and I was confused with the inversion of operands.